emar.py

```diff
@@ -23,6 +24,19 @@ def run():
   if DEBUG == "0":
     DEBUG = None
 
+  # read response files very early on
+  response_file = True
+  while response_file:
+    response_file = None
+    for index in range(1, len(sys.argv)):
+      if sys.argv[index][0] == '@':
+        # found one, loop again next time
+        response_file = True
+        extra_args = read_response_file(sys.argv[index])
+        # slice in extra_args in place of the response file arg
+        sys.argv[index:index+1] = extra_args
```

Comment:
This would be good to go also in `response_file.py` as a function such as `expand_response_files(cmdline)`, and here we could call `sys.argv = expand_response_files(sys.argv)` - that way other parts of the toolchain will have this ability to recursively expand response files available to it.

Also, I think since this inserts `extra_args` inline, the whole `response_file = None/True` and double loop logic can be removed with a single forward rolling fill, with something like

```python
def expand_response_files(cmd_line):
  cmd_line = cmd_line[:] # Don't mutate the original list
  index = 0
  while index < len(cmd_line):
    if cmd_line[index][0] == '@':
      cmd_line[index:index+1] = read_response_file(cmd_line[index])
    else:
      index += 1
  return cmd_line
```

that way it will be of linear complexity instead of quadratic, and perhaps a bit easier to parse?

Reply:
I agree...
But this is exactly how read_response_file is used in emcc.py https://github.com/kripken/emscripten/blob/incoming/emcc.py#L318-L329

Reply:
Those lines could then become `sys.argv = expand_response_files(sys.argv)` as well, and the implementation would be shared?

Reply:
yes